proxy/src/avatar.rs

```diff
@@ -0,0 +1,201 @@
+//! Org and user avatar generation.
+
+/// Emoji whitelist.
+const EMOJIS: &[&str] = &[
+    "⌚️", "📱", "📲", "💻", "⌨️", "💽", "💾", "💿", "📀", "📼", "📷", "📸", "📹", "🎥", "🎞", "📞",
+    "☎️", "📟", "📠", "📺", "📻", "⏰", "🕰", "⌛️", "⏳", "📡", "🔋", "🔌", "💡", "🔦", "💸", "💵",
+    "💴", "💶", "💷", "💰", "💳", "🧾", "💎", "⚖️", "🔧", "🔨", "", "🔩", "⚙️", "⛓", "🔫", "💣", "🔪",
+    "🗡", "⚔️", "🚬", "⚰️", "⚱️", "🏺", "🧭", "🧱", "🔮", "🧿", "🧸", "📿", "💈", "⚗️", "🔭", "🧰", "🧲", "🧪",
+    "🧫", "🧬", "🧯", "🔬", "🕳", "💊", "💉", "🚽", "🚰", "🚿", "🛁", "🛀", "🛀🏻", "🛀🏼", "🛀🏽",
+    "🛀🏾", "🛀🏿", "🧴", "🧵", "🧶", "🧷", "🧹", "🧺", "🧻", "🧼", "🧽", "🛎", "🔑", "🗝", "🚪", "🛍", "🧳",
+    "🛒", "🎁", "🎈", "🎏", "🎀", "🎊", "🎉", "🧨", "🎎", "🏮", "🎐", "🧧", "✉️", "📩", "📨", "📧",
+    "💌", "📥", "📤", "📦", "📪", "📫", "📬", "📭", "📮", "📯", "📜", "📃", "📄", "📑", "📊", "📈",
+    "📉", "📆", "📅", "📇", "📋", "📁", "📂", "🗂", "🗞", "📰", "📓", "📔", "📒", "📕", "📗", "📘",
+    "📙", "📚", "📖", "🔖", "🔗", "📎", "📐", "📏", "📌", "📍", "✂️", "🖊", "🖋", "✒️", "🖌", "📝", "✏️",
+    "🔍", "🔎", "🔏", "🔐",
+];
+
+/// An emoji.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Emoji(&'static str);
+
+/// An avatar.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Avatar {
+    /// The emoji component.
+    pub emoji: Emoji,
+    /// The background color component.
+    pub background: Color,
+}
+
+impl Avatar {
+    /// Generate an avatar from an input string.
+    pub fn from(input: &str) -> Self {
+        Self {
+            emoji: generate_emoji(input),
+            background: compress_color(generate_color(input)),
+        }
+    }
+}
+
+/// A 32-bit RGBA color.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Color {
+    /// The red channel.
+    pub r: u8,
+    /// The green channel.
+    pub g: u8,
+    /// The blue channel.
+    pub b: u8,
+
+    // The alpha is here to facilitate working with `u32` values.
+    // We don't use it as part of the output.
+    a: u8,
+}
+
+impl Color {
+    /// Create a new color from individual channels.
+    pub fn new(r: u8, g: u8, b: u8) -> Self {
+        Self { r, g, b, a: 0x0 }
+    }
+
+    /// Compute the lightness of a color.
+    pub fn lightness(&self) -> f32 {
+        let r = self.r as f32;
+        let g = self.g as f32;
+        let b = self.b as f32;
+        let n = 0xff as f32;
+
+        // This isn't perceptual lightness, but whatever.
+        (r / n + g / n + b / n) / 3.
+    }
+
+    /// Ligthen a color by an amount between `-1.0` and `1.0`.
+    /// Panics if the amount isn't in that range.
+    fn lighten(self, amount: f32) -> Self {
+        assert!(amount <= 1.0);
+        assert!(amount >= -1.0);
```

Comment:
What does the compiler do to these macros when built with `--releae`? Generally if this is fallible I'd prefer it to return a `Result` to avoid panics in library code.

Reply:
These stay around in release builds. You can use `assert_debug!` for debug-only asserts. The function isn't failible, but it can mis-used. If you don't like this approach, I can also just cap the amount, but this can hide bugs.

Reply:
e34d612

Reply:
On the fence, what I don't want is panics in library code. Esp. from something like a color manipulation input being out of bound. Ideally the input given can never be incorrect, but that might go out of hand quickly for what we want to do. So either we cap it or return a `Result`, we have a already good amount of error infrastrcture in `error.rs`.